go/viperutil/viper.go

```diff
@@ -0,0 +1,186 @@
+package viperutil
+
+import (
+	"fmt"
+
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+)
+
+type Option[T any] func(val *Value[T])
+
+func WithAliases[T any](aliases ...string) Option[T] {
+	return func(val *Value[T]) {
+		val.aliases = append(val.aliases, aliases...)
+	}
+}
+
+func WithDefault[T any](t T) Option[T] {
+	return func(val *Value[T]) {
+		val.hasDefault, val.value = true, t
+	}
+}
+
+func WithEnvVars[T any](envvars ...string) Option[T] {
+	return func(val *Value[T]) {
+		val.envvars = append(val.envvars, envvars...)
+	}
+}
+
+func WithFlags[T any](flags ...string) Option[T] {
+	return func(val *Value[T]) {
+		val.flagNames = append(val.flagNames, flags...)
+	}
+}
+
+type Value[T any] struct {
+	key   string
+	value T
+
+	aliases    []string
+	flagNames  []string
+	envvars    []string
+	hasDefault bool
+
+	resolve func(string) T
+	loaded  bool
+}
+
+// NewValue returns a viper-backed value with the given key, lookup function,
+// and bind options.
+func NewValue[T any](key string, getFunc func(string) T, opts ...Option[T]) *Value[T] {
+	val := &Value[T]{
+		key:     key,
+		resolve: getFunc,
+	}
+
+	for _, opt := range opts {
+		opt(val)
+	}
+
+	return val
+}
+
+// Bind binds the value to flags, aliases, and envvars, depending on the Options
+// the value was created with.
+//
+// If the passed-in viper is nil, the global viper instance is used.
+//
+// If the passed-in flag set is nil, flag binding is skipped. Otherwise, if any
+// flag names do not match the value's canonical key, they are registered as
+// additional aliases for the value's key. This function panics if a flag name
+// is specified that is not defined on the flag set.
+func (val *Value[T]) Bind(v *viper.Viper, fs *pflag.FlagSet) {
+	if v == nil {
+		v = viper.GetViper()
+	}
+
+	aliases := val.aliases
+
+	// Bind any flags, additionally aliasing flag names to the canonical key for
+	// this value.
+	if fs != nil {
+		for _, name := range val.flagNames {
+			f := fs.Lookup(name)
+			if f == nil {
+				// TODO: Don't love "configuration error" here as it might make
+				// users think _they_ made a config error, but really it is us,
+				// the Vitess authors, that have misconfigured a flag binding in
+				// the source code somewhere.
+				panic(fmt.Sprintf("configuration error: attempted to bind %s to unspecified flag %s", val.key, name))
+			}
+
+			if f.Name != val.key {
+				aliases = append(aliases, f.Name)
+			}
+
+			// We are deliberately ignoring the error value here because it
+			// only occurs when `f == nil` and we check for that ourselves.
+			_ = v.BindPFlag(val.key, f)
+		}
+	}
+
+	// Bind any aliases.
+	for _, alias := range aliases {
+		v.RegisterAlias(alias, val.key)
+	}
+
+	// Bind any envvars.
+	if len(val.envvars) > 0 {
+		vars := append([]string{val.key}, val.envvars...)
+		// Again, ignoring the error value here, which is non-nil only when
+		// `len(vars) == 0`.
+		_ = v.BindEnv(vars...)
```

Comment:
`BindEnv` is an outlier, in that if you:

1. `v.BindEnv()` => returns error
2. `v.BindEnv(key)` => binds the environment variable `strings.ToUpper(strings.ReplaceAll(key, ".", "_"))`
3. `v.BindEnv(key, var1, var2, ...)` => does (2) and also binds to whatever the literal strings contained in (`var1`, `var2, ...) are

The case in (2) is particularly of interest; if we didn't do this slice creation here, then regardless of whether a value was specified with a `WithEnvVars` option, we would bind to the full key name but uppercased and `s/./_`. This would mean that any value _whatsoever_ gets bound to its envvar-ified key name with no way to disable or opt-out of that behavior.

Reply:
(I would have written this in a comment before taking out of draft, so sorry about that! I was looking for the feel/devex on the API for writing viper flags at this stage)

Reply:
Ah, I see. Thanks! And no worries, I was just curious. :-)